### ISSUE TITLE
Facing issues with Low Rank Transforms 

### ISSUE THREAD

**Issue by SourabhRanade** [amandeepsp/amandeepsp.github.io#49]
Hello @amandeepsp , I was referring to your blog at [https://amandeepsp.github.io/making-models-smaller-2](url) and trying to make it work on a Mask RCNN with a SWIN transformer. I am using a pretrained model for this purpose and using mmdetection for the same. 
However, when I do the transformation on the model, I see that the model is not able to detect anything. It simply returns the image that I pass it as an input. Just the pretrained model without the below code or the pretrained model quantized using dynamic quantization works properly.  I was wondering if maybe you could help me out or guide me to the possible error in this.

The following is the code with which I load the model after I have downloaded them from the appropriate place
```

!wget -c https://download.openmmlab.com/mmdetection/v2.0/swin/mask_rcnn_swin-t-p4-w7_fpn_1x_coco/mask_rcnn_swin-t-p4-w7_fpn_1x_coco_20210902_120937-9d6b7cfa.pth \ -O checkpoints/mask_rcnn_swin-t-p4-w7_fpn_1x_coco_20210902_120937-9d6b7cfa.pth


from mmdet.apis import init_detector, inference_detector

config='/content/mmdetection/configs/swin/mask_rcnn_swin-t-p4-w7_fpn_1x_coco.py'
# Setup a checkpoint file to load
checkpoint = 'checkpoints/mask_rcnn_swin-t-p4-w7_fpn_1x_coco_20210902_120937-9d6b7cfa.pth'
# initialize the detector
model = init_detector(config,checkpoint)
```

The following is the code that I use for the low rank transformation:

```
def replace_layers(model):
    svd_classifier_layers=[]
    L=40
    count=1
    for module_name, module in model.named_children():
      if module_name=='proj':
        continue
      if len(list(module.children())) > 0:
        replace_layers(module)
      if isinstance(module, torch.nn.Linear):
        W=module.weight.data
        # print(W)
        # U,S,V=torch.svd(W)
        U, S, Vh = torch.linalg.svd(W)
        V = Vh.transpose(-2, -1).conj()
        W1 = U[:,:L]
        W2 = torch.diag(S[:L]) @ V[:,:L].t()
        if module.bias is not None:
          # print("W1:",W1)
          layer_1 = nn.Linear(in_features=module.in_features,
                            out_features=L, bias=True)
          layer_1.weight.data = W2
          svd_classifier_layers.append(layer_1)

          layer_2 = nn.Linear(in_features=L,
                            out_features=module.out_features , bias=True)
          layer_2.weight.data = W1
          layer_2.bias.data = module.bias.data
          svd_classifier_layers.append(layer_2)
          # print("Layer2",layer_2.weight.data)
          # break
        else:
          layer_1 = nn.Linear(in_features=module.in_features,
                            out_features=L, bias=False)
          layer_1.weight.data = W2
          svd_classifier_layers.append(layer_1)

          layer_2 = nn.Linear(in_features=L,
                            out_features=module.out_features , bias=False)
          layer_2.weight.data = W1
          # layer_2.bias.data = module.bias.data
          svd_classifier_layers.append(layer_2)

        try:
          # print("we are now trying to replace",module,"with",svd_classifier_layers,"in a sequnetial way")
          # n =int(module_name)
          # model[n]=nn.Sequential(*svd_classifier_layers)
          setattr(model, module_name, nn.Sequential(*svd_classifier_layers))
        except:
          setattr(model, module_name, nn.Sequential(*svd_classifier_layers))
```

Currently, I am only passing the model's backbone to the above function. 

`
replace_layers(model.backbone)
`

The model size does reduce a bit with this but I am not sure why the inference done through mmdetection won't work. It would be great if you could guide me to the right place

**Comment by amandeepsp** [comment]
Low Rank Transformations are only viable for models with high Fully connected to Conv weight ratio, which is not true for many modern model architectures (also this case). Even if you could make the detection work, the size decrease wouldn't be worth it. If you need a smaller model, please consider using a MobileNet backbone here.
